src/game/map.rs

```diff
@@ -106,80 +105,84 @@ pub fn find_exit(from_room: RoomName, to_room: RoomName) -> Result<ExitDirection
-        let code: i32 = js_unwrap! {Game.map.findExit(@{from_room}, @{to_room}, @{callback})};
-        ExitDirection::from_i32(code)
-            .map(Ok)
-            .or_else(|| ReturnCode::from_i32(code).map(Err))
-            .unwrap_or_else(|| {
-                panic!(
-                    "find_exit: return value {:?} not recognized as either Exit nor ReturnCode",
-                    code
-                )
-            })
-    })
+    let code: i32 = js!(
+        let cb = @{callback_lifetime_erased};
+        let res = Game.map.findExit(@{from_room}, @{to_room}, cb);
+        cb.drop();
+        return res;
+    )
+    .try_into()
+    .expect("expected int from findExit");
+
+    ExitDirection::from_i32(code)
+        .map(Ok)
+        .or_else(|| ReturnCode::from_i32(code).map(Err))
+        .unwrap_or_else(|| {
+            panic!(
+                "find_exit: return value {:?} not recognized as either Exit nor ReturnCode",
+                code
+            )
+        })
 }
 
 pub fn find_route(from_room: &str, to_room: &str) -> Result<Vec<RoomRouteStep>, ReturnCode> {
     let v = js!(return Game.map.findRoute(@{from_room}, @{to_room}););
     parse_find_route_returned_value(v)
 }
 
-scoped_thread_local!(static FR_CALLBACK: Box<(dyn Fn(RoomName, RoomName) -> f64)>);
-
 pub fn find_route_with_callback(
     from_room: RoomName,
     to_room: RoomName,
-    route_callback: impl Fn(RoomName, RoomName) -> f64,
+    route_callback: impl FnMut(RoomName, RoomName) -> f64,
 ) -> Result<Vec<RoomRouteStep>, ReturnCode> {
-    // Actual callback
-    fn callback(room_name: String, from_room_name: String) -> f64 {
-        FR_CALLBACK.with(|callback| {
-            callback(
-                room_name.parse().expect(
-                    "expected room name passed into Game.map.findRoute \
-                     callback to be a valid room name",
-                ),
-                from_room_name.parse().expect(
-                    "expected room name passed into Game.map.findRoute \
-                     callback to be a valid room name",
-                ),
-            )
-        })
-    }
-
-    let callback_type_erased: Box<dyn Fn(RoomName, RoomName) -> f64> = Box::new(route_callback);
-
-    let callback_lifetime_erased: Box<dyn Fn(RoomName, RoomName) -> f64 + 'static> =
+    let mut raw_callback = route_callback;
+
+    let mut callback_boxed = move |to_name: RoomName, from_name: RoomName| -> f64 {
+        raw_callback(to_name, from_name).into()
+    };
+
+    // Type erased and boxed callback: no longer a type specific to the closure
+    // passed in, now unified as &Fn
+    let callback_type_erased: &mut (dyn FnMut(RoomName, RoomName) -> f64) =
+        &mut callback_boxed;
+
+    // Overwrite lifetime of reference so it can be passed to javascript. 
+    // It's now pretending to be static data. This should be entirely safe
+    // because we're only sticking it in scoped storage and we control the
```

Comment:
Might be worth removing the reference to scoped storage?

I think the point still stands, it's just that we're trusting `Game.map.findRoute` to never persist our closure, rather than trusting `scoped_tls`.